mmpretrain/models/backbones/hivit.py

```diff
@@ -0,0 +1,609 @@
+
+        # stochastic depth
+        dpr = iter(x.item() for x in torch.linspace(0, drop_path_rate, sum(depths) + sum(depths[:-1])))
+
+        # build blocks
+        self.blocks = nn.ModuleList()
+        for stage_i, stage_depth in enumerate(depths):
+            is_main_stage = embed_dim == self.num_features
+            nhead = num_heads if is_main_stage else 0
+            ratio = mlp_ratio if is_main_stage else stem_mlp_ratio
+            # every block not in main stage includes two mlp blocks
+            stage_depth = stage_depth if is_main_stage else stage_depth * 2
+            for _ in range(stage_depth):
+                self.blocks.append(
+                    BlockWithRPE(
+                        Hp, embed_dim, nhead, ratio, qkv_bias, qk_scale,
+                        drop=drop_rate, attn_drop=attn_drop_rate, drop_path=next(dpr), rpe=rpe,
+                        norm_cfg=norm_cfg, layer_scale_init_value=layer_scale_init_value,
+                    )
+                )
+            if stage_i + 1 < self.num_stages:
+                self.blocks.append(
+                    PatchMerge(embed_dim, norm_cfg)
+                )
+                embed_dim *= 2
+
+        self.frozen_stages = frozen_stages
+        if self.frozen_stages > 0:
+            self._freeze_stages()
+
+        self.apply(self._init_weights)
+
+    def _init_weights(self, m):
+        if isinstance(m, nn.Linear):
+            trunc_normal_(m.weight, std=.02)
+            if isinstance(m, nn.Linear) and m.bias is not None:
+                nn.init.constant_(m.bias, 0)
+        elif isinstance(m, nn.LayerNorm):
+            nn.init.constant_(m.bias, 0)
+            nn.init.constant_(m.weight, 1.0)
+
+    @torch.jit.ignore
+    def no_weight_decay(self):
+        return {'pos_embed'}
+
+    @torch.jit.ignore
+    def no_weight_decay_keywords(self):
+        return {'relative_position_bias_table'}
+
+    def interpolate_pos_encoding(self, x, h, w):
+        npatch = x.shape[1]
+        N = self.pos_embed.shape[1]
+        if npatch == N and w == h:
+            return self.pos_embed
+        patch_pos_embed = self.pos_embed
+        dim = x.shape[-1]
+        w0 = w // self.patch_size
+        h0 = h // self.patch_size
+        # we add a small number to avoid floating point error in the interpolation
+        # see discussion at https://github.com/facebookresearch/dino/issues/8
+        w0, h0 = w0 + 0.1, h0 + 0.1
+        patch_pos_embed = nn.functional.interpolate(
+            patch_pos_embed.reshape(1, int(math.sqrt(N)), int(math.sqrt(N)), dim).permute(0, 3, 1, 2),
+            scale_factor=(h0 / math.sqrt(N), w0 / math.sqrt(N)),
+            mode='bicubic',
+        )
+        assert int(h0) == patch_pos_embed.shape[-2] and int(w0) == patch_pos_embed.shape[-1]
+        patch_pos_embed = patch_pos_embed.permute(0, 2, 3, 1).view(1, -1, dim)
+        return patch_pos_embed
+
+    def forward_features(self, x, mask=None):
```

Comment:
the code of `forward_features` can be moved to forward for simplicity

Reply:
forward_features is needed for self-supervised pre-training